attr/type.go

```diff
@@ -92,6 +92,15 @@ type TypeWithElementTypes interface {
 	ElementTypes() []Type
 }
 
+// TypeWithDynamicValue extends the Type interface to include information about how
+// a dynamic type will determine the type of a dynamic value.
+type TypeWithDynamicValue interface {
+	Type
+
+	// DetermineAttrType returns a corresponding attr.Type, based on the tftypes.Type representation.
+	DetermineAttrType(tftypes.Type) Type
```

Comment:
Thanks for pushing this, I believe I have this resolved now! (removed the `attr.TypeWithDynamicValue` completely and switched all detection logic to use `basetypes.DynamicTypable`) c920b8a ⚡  

> To bridge the gap I was having to remove this earlier, when we were switching the function parameters to use `types.Tuple` for variadic arguments, I half implemented the reflection support for dynamic values believing that we would eventually want that.
>
> We've since decided that we are going to wait to decide if we want to implement any reflection rules, but if we do, it'll likely be a defined table of rules. The reason `tftypes.Type` was used in the interface is because our reflection logic currently relies on it 😃 
>

I removed the reflection support that was implemented and replaced it with a nice error message suggesting to use the actual `types` package, which is supported 🎸

Reply:
That's awesome and https://github.com/hashicorp/terraform-plugin-framework/commit/c920b8a108530a0a52f5e65ba12a430c0eac1c0f looks great, thank you!